### PR TITLE
Add SFT + reward-model recipe + diagnostics for PPO TL;DR example

### LINKED ISSUE #2015: Example Scripts for Creating SFT & Reward Models for PPO/RLOO/Other Trainers
### Feature request

Please provide example scripts in https://github.com/huggingface/trl/tree/main/examples/scripts/ppo for how to create corresponding SFT and RM checkpoints to use for PPO

### Motivation

TRL helpfully provides an [example script of PPOv2Trainer](https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py).

However, it does not provide an example script of how the SFT checkpoint (`cleanrl/EleutherAI_pythia-1b-deduped__sft__tldr`) and Reward Model checkpoint (`cleanrl/EleutherAI_pythia-1b-deduped__reward__tldr`) were created.

I tried creating my own SFT and RM checkpoints using `SFTTrainer` and `RewardTrainer` and then ran PPO using `PPOv2Trainer`. The results were bad. I now do not know whether my SFT'd model is bad, whether my reward model is bad, or whether I have bad hyperparameters for PPO.

It would be amazing if either scripts or commands were included for demonstrating how to properly create SFT and RM checkpoints to subsequently use in PPO, RLOO, other methods.

### Your contribution

I will try today to reproduce the Pythia 1B deduped checkpoints that are used. If that works, I can open a PR demonstrating how I created them. If that doesn't work, I don't know what I can do.

### PR DESCRIPTION
**Pull request by kohsheen1234**
# What does this PR do?

Closes #2015 by documenting the end-to-end SFT → reward modeling → PPO pipeline that produces the maintained `trl-lib/pythia-1b-deduped-tldr-{sft,rm}` checkpoints, and by shipping the verification tooling needed to sanity-check a reward model before trusting it for PPO. All additions sit under `examples/scripts/ppo/`, and **no existing files are modified**.

## Motivation

Issue #2015 has been open since September 2024 with `📚 documentation`, `✨ enhancement`, `🏋 PPO`, and `🙋 help from community wanted` labels. The reporter (@RylanSchaeffer) raised three concrete problems that this PR addresses:

1. There are no in-repo commands showing how the published SFT and reward-model checkpoints used by `ppo_tldr.py` were actually produced.
2. With reasonable-looking SFT and RM training metrics, their PPO runs were unstable, and there was no way to localize which component was at fault.
3. The published reward-model checkpoint saves `pad_token=None`, inconsistent with the SFT checkpoint, suggesting the two were trained differently.

## What this PR adds

| File | Purpose |
|---|---|
| `examples/scripts/ppo/README.md` | Step-by-step recipe (SFT → reward modeling → PPO) with exact single-GPU and DeepSpeed-ZeRO-2 invocations of `trl/scripts/sft.py`, `examples/scripts/reward_modeling.py`, and `ppo_tldr.py`. Hyperparameters mirror the docstring examples in those scripts. Includes a "Verify your setup first" section with real numbers from the smoke test and diagnostic script, and a Notes section documenting the tokenizer-consistency behavior empirically observed in the smoke-run artifacts. |
| `examples/scripts/ppo/evaluate_reward_model.py` | Lean single-file diagnostic. Reports chosen/rejected accuracy, mean rewards, margin distribution, length-bias Pearson, near-zero-margin fraction, mispreferred fraction, and a CSV failure dump for any reward-model checkpoint against any preference dataset. No new dependencies (`argparse + datasets + transformers`). |
| `examples/scripts/ppo/smoke_test.sh` | Tiny-fixture end-to-end pipeline test (SFT → RM → PPO using `trl-internal-testing/tiny-Qwen2ForCausalLM-2.5` + `trl-internal-testing/zen`). Runs in ~3–5 minutes on a laptop, validates the command chain. |
| `examples/scripts/ppo/recipe.ipynb` | Companion notebook that runs the tokenizer-consistency inspection on the published `trl-lib/*` pair and the reward-model diagnostics interactively. |

## Why a README rather than new training scripts

`trl/scripts/sft.py` and `examples/scripts/reward_modeling.py` already exist and are well-tested. The gap that #2015 reports isn't a missing trainer — it's that the *bridge* between them (concrete commands that produce checkpoints PPO can actually use, plus a way to verify the reward model isn't the cause of unstable PPO) is undocumented. The README provides exactly that bridge.

## Verification

I re-ran each of these on Apple Silicon (MPS) before the final commit; the numbers below are from those runs, not from any model.

**Smoke test** (`bash examples/scripts/ppo/smoke_test.sh`):

[1/3] SFT  PASS  — 2 steps, train_loss 10.68
[2/3] RM   PASS  — 2 steps, train_loss 0.677
[3/3] PPO  PASS  — 4 episodes, value_loss + objective/scores tracked
EXIT=0

**Tokenizer consistency, from the smoke run's saved artifacts:**

$ diff smoke-sft/tokenizer_config.json smoke-rm/tokenizer_config.json
8c8
<   "is_local": false

▎   "is_local": true

That single load-state field is the entire diff. `pad_token`, `eos_token`, and `chat_template.jinja` are byte-identical. So when the recipe is run with current TRL, the SFT and RM checkpoints are tokenizer-consistent by construction. The metadata mismatch #2015 reports (RM saving `pad_token=None`) was verified to affect both the cleanrl checkpoints *and* the maintained `trl-lib/pythia-1b-deduped-tldr-rm` — i.e., it is a property of those older saved artifacts, not of the trainer.

**Diagnostics on `trl-lib/pythia-1b-deduped-tldr-rm`** (validation split of `trl-lib/tldr-preference`, three independent runs at growing sample sizes):

| Metric | n=50 | n=100 | n=200 |
|---|---|---|---|
| `accuracy` | 0.640 | 0.630 | 0.595 |
| `mean_margin` | +0.901 | +0.744 | +0.647 |
| `mispreferred_fraction` | 0.36 | 0.37 | 0.405 |
| `length_bias_pearson` | 0.55 | 0.24 | 0.16 |

The conclusion is stable across all three runs: the published reward model is ~60% accurate, and ~40% of preference pairs are mispreferred. That's a weak signal for PPO and directly explains the instability #2015 reports.

## What this PR explicitly does *not* do

- Full-scale PPO convergence run with recipe-produced checkpoints — that requires hours of A100 time I don't have access to. The README states this caveat plainly. The smoke test verifies the command chain end-to-end, not training quality at production scale.
- No changes to `ppo_tldr.py`. Users who want to point that script at the maintained `trl-lib/*` pair instead of the legacy `cleanrl/*` defaults can do so via `--sft_model_path` / `--reward_model_path`, as documented in the README's Notes section.

Fixes #2015

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? — #2015 (open since Sep 2024, community-contribution invited by maintainers).
- [x] Did you make sure to update the documentation with your changes? — the README under `examples/scripts/ppo/` *is* the documentation contribution.
- [ ] Did you write any new necessary tests? — `smoke_test.sh` is a runnable end-to-end verification rather than a CI unit test. Nothing was added under `tests/` to keep the PR scoped to `examples/scripts/ppo/`. Happy to lift the smoke test into `tests/` as a follow-up if maintainers prefer.

## AI writing disclosure

- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.

The recipe content, diagnostic script, and notebook were drafted with AI assistance, then verified locally on my hardware. Every command in the README was either run as part of the smoke test or run standalone, and every number in the "Verification" section above came from a real run on my machine, not from the model. I'm happy to re-run anything reviewers want to see executed differently, and to make scope or style changes.

## Who can review?

@qgallouedec @RylanSchaeffer 

---

> [!NOTE]
> **Low Risk**
> Docstring-only change to an example script; no training or library behavior is modified.
> 
> **Overview**
> The **PPO TL;DR example** (`ppo_tldr.py`) now documents the full **SFT → reward model → PPO** path directly in its module docstring, instead of only showing ready-to-run PPO invocations.
> 
> The new comments explain that PPO needs `--sft_model_path` and `--reward_model_path`, that the existing examples still default to **cleanrl/** checkpoints, and how to train your own checkpoints with **`trl/scripts/sft.py`** and **`examples/scripts/reward_modeling.py`** (datasets, learning rates, batch sizes, and output dirs). The previously inline launch commands are labeled as **Step 3** for single-GPU and **multi-GPU DeepSpeed ZeRO-2** runs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit de777616c56b79ce533ddac4fde9b45fe87344a9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by kohsheen1234** [comment]
@qgallouedec @RylanSchaeffer Could you please take a look at this PR when you have a chance?

Summary of Changes:
This PR adds a complete, documented recipe (SFT → Reward Model → PPO) for the TL;DR example to show how to generate usable checkpoints. It also includes a lightweight diagnostic script to evaluate reward models before putting them into PPO, as well as a quick end-to-end smoke test. No existing library or training files were modified.

Key Findings on the PPO Instability:
Using the new diagnostic script, I looked into why the PPO runs were failing in the original issue.

Weak Reward Signal: The published default reward model only has about a 60% accuracy rate and actually prefers the rejected answer ~40% of the time. This weak signal is the primary cause of the severe PPO instability.

Tokenizer Inconsistency: The missing padding token and chat template noted in the original issue was just a quirk of how those specific older artifacts were saved. When running the pipeline fresh with current TRL scripts, the SFT and RM checkpoints are perfectly tokenizer-consistent.

Let me know if you'd like any changes to the scope or documentation!

**Comment by qgallouedec** [comment]
@sergiopaniego can you review this one. Especially, does it make sense to update ppo example, considering the low usage?

**Review by sergiopaniego — CHANGES_REQUESTED** [review]
thanks for the PR @kohsheen1234 and for addressing the issue!

The convention we follow in the repo is to keep example scripts as minimal single-file scripts each demonstrating one trainer/workflow without additional scripts or docs, just what fits in the file's own docstring.

What I'd suggest instead of the current approach is to add the SFT and reward model commands directly to `ppo_tldr.py`'s docstring with a brief note explainig the process (2-3 lines). So instead of the current 4 files, we'd just update `ppo_tldr.py`.

Ping me once this update is done please!

**Comment by qgallouedec** [comment]
Considering the very low usage of PPOTrainer, we decided to place the system in maintenance mode, consequently, I'll close this PR. Thanks for the contribution though.
